whisper.cpp

```diff
@@ -2817,12 +2771,59 @@ struct whisper_state * whisper_init_state(whisper_context * ctx) {
     state->decoders[0].probs.reserve(ctx->vocab.n_vocab);
     state->decoders[0].logits.reserve(ctx->vocab.n_vocab);
     state->decoders[0].logprobs.reserve(ctx->vocab.n_vocab);
-    state->buf_compute.resize(scale * std::max(MEM_REQ_ENCODE.at(ctx->model.type), MEM_REQ_DECODE.at(ctx->model.type)));
 
-    state->buf_scratch[0].resize(MEM_REQ_SCRATCH0.at(ctx->model.type));
-    state->buf_scratch[1].resize(MEM_REQ_SCRATCH1.at(ctx->model.type));
-    state->buf_scratch[2].resize(MEM_REQ_SCRATCH2.at(ctx->model.type));
-    state->buf_scratch[3].resize(MEM_REQ_SCRATCH3.at(ctx->model.type));
+    state->buf_compute.resize(ggml_tensor_overhead()*GGML_MAX_NODES + ggml_graph_overhead());
+
+    static const size_t tensor_alignment = 32;
+
+    state->alloc_encode      = ggml_allocr_new_measure(tensor_alignment);
+    state->alloc_encode_post = ggml_allocr_new_measure(tensor_alignment);
+    state->alloc_decode      = ggml_allocr_new_measure(tensor_alignment);
```

Comment:
There is a chance that this will not work in some systems with limited virtual memory, such as iOS, because each measure allocator reserves a large amount of virtual memory. It would be safer to allocate only one measure allocator at a time, I think that should be possible here.
It's definitely not ideal that ggml-alloc has this limitation, I expect to improve this and remove the use of virtual memory entirely with the common backends interface implementation.

Reply:
I think I reorganized the allocators as proposed, but it seems some OSes still fail - see `linux/arm64` and `linux/ppc64le` in the CI: https://github.com/ggerganov/whisper.cpp/actions/runs/6146272809/job/16675319319

Reply:
I am a bit confused by this, it seems that the call to `mmap` is crashing the process instead of returning an error, because otherwise we should see the failed assert `GGML_ASSERT(!"failed to allocate virtual memory for measure buffer");`. I imagine that this is related to QEMU, I'll try to reproduce it locally.

Reply:
I tried the exact same commands that the CI uses to run the arm64 version with docker and QEMU, and it works on my computer. So whatever is the issue, it only seems to happen on the github CI environment and I cannot reproduce it. Maybe it is hitting some memory usage limit.

<details>

```bash
$ sudo docker run --platform linux/arm64 --rm \
    -v /home/diego/code/whisper.cpp:/workspace \
    -w /workspace ubuntu:22.04 /bin/sh -c '
    apt update
    apt install -y build-essential cmake libsdl2-dev
    cmake . -DWHISPER_SUPPORT_SDL2=ON -DCMAKE_BUILD_TYPE=Release
    make
    ctest -L gh --output-on-failure'

[...]

-- The C compiler identification is GNU 11.4.0
-- The CXX compiler identification is GNU 11.4.0
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Check for working C compiler: /usr/bin/cc - skipped
-- Detecting C compile features
-- Detecting C compile features - done
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Check for working CXX compiler: /usr/bin/c++ - skipped
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Could NOT find Git (missing: GIT_EXECUTABLE)
-- Looking for pthread.h
-- Looking for pthread.h - found
-- Performing Test CMAKE_HAVE_LIBC_PTHREAD
-- Performing Test CMAKE_HAVE_LIBC_PTHREAD - Success
-- Found Threads: TRUE
-- CMAKE_SYSTEM_PROCESSOR: aarch64
-- ARM detected
-- Configuring done
-- Generating done
CMake Warning:
  Manually-specified variables were not used by the project:

    WHISPER_SUPPORT_SDL2


-- Build files have been written to: /workspace
[  7%] Building C object CMakeFiles/whisper.dir/ggml.c.o
[ 15%] Building C object CMakeFiles/whisper.dir/ggml-alloc.c.o
[ 23%] Building CXX object CMakeFiles/whisper.dir/whisper.cpp.o
[ 30%] Linking CXX shared library libwhisper.so
[ 30%] Built target whisper
[ 38%] Building CXX object examples/CMakeFiles/common.dir/common.cpp.o
[ 46%] Building CXX object examples/CMakeFiles/common.dir/common-ggml.cpp.o
[ 53%] Linking CXX static library libcommon.a
[ 53%] Built target common
[ 61%] Building CXX object examples/main/CMakeFiles/main.dir/main.cpp.o
[ 69%] Linking CXX executable ../../bin/main
[ 69%] Built target main
[ 76%] Building CXX object examples/bench/CMakeFiles/bench.dir/bench.cpp.o
[ 84%] Linking CXX executable ../../bin/bench
[ 84%] Built target bench
[ 92%] Building CXX object examples/quantize/CMakeFiles/quantize.dir/quantize.cpp.o
[100%] Linking CXX executable ../../bin/quantize
[100%] Built target quantize
Test project /workspace
    Start 1: test-main-tiny
1/2 Test #1: test-main-tiny ...................   Passed   82.79 sec
    Start 2: test-main-tiny.en
2/2 Test #2: test-main-tiny.en ................   Passed   83.46 sec

100% tests passed, 0 tests failed out of 2

Label Time Summary:
en      =  83.46 sec*proc (1 test)
gh      = 166.24 sec*proc (2 tests)
tiny    = 166.24 sec*proc (2 tests)
```

</details>

Reply:
A possible workaround could be reducing the amount of virtual memory allocated here:
https://github.com/ggerganov/whisper.cpp/blob/d3b2dd4955e40172c2dec4a39a3f0c2874a905be/ggml-alloc.c#L345-L346

Reply:
Ok, thanks for looking into this. I'll now continue working on this branch and try to find a solution

Reply:
Reducing the size to 128GB fixes the CI: https://github.com/ggerganov/whisper.cpp/pull/1270/commits/b19888cfb48d135466a890bccb348ad4e0772b14